components/brave_wallet_ui/components/extension/swap/index.tsx

```diff
@@ -0,0 +1,349 @@
+} from '../../../common/slices/api.slice'
+
+type SwapToken = Pick<
+  BraveWallet.BlockchainToken,
+  | 'chainId'
+  | 'coin'
+  | 'symbol'
+  | 'contractAddress'
+  | 'logo'
+  | 'isErc721'
+  | 'isNft'
+  | 'name'
+  | 'coingeckoId'
+  | 'decimals'
+>
+
+const makeToken = (
+  token?: SwapToken,
+  network?: BraveWallet.NetworkInfo,
+  symbol?: string,
+  decimals?: number
+) => {
+  if (!token || !network) {
+    return undefined
+  }
+
+  if (token.coingeckoId !== UNKNOWN_TOKEN_COINGECKO_ID) {
+    return token
+  }
+
+  if (token.contractAddress === '') {
+    return {
+      ...token,
+      decimals: network.decimals,
+      symbol: network.symbol
+    }
+  }
+
+  if (!symbol || !decimals) {
+    return undefined
+  }
+
+  return {
+    ...token,
+    symbol,
+    decimals
+  }
+}
+
+
+interface Props {
+  sellToken?: SwapToken
+  buyToken?: SwapToken
+  sellAmount?: string
+  buyAmount?: string
+  senderOrb: string
+  senderLabel?: string
+  recipientOrb?: string
+  recipientLabel?: string
+  expectRecipientAddress?: boolean
+}
+
+export function SwapBase(props: Props) {
+  const {
+    sellToken,
+    buyToken,
+    sellAmount,
+    buyAmount,
+    senderLabel,
+    senderOrb,
+    recipientOrb,
+    recipientLabel,
+    expectRecipientAddress
+  } = props
+
+  // queries
+  const { data: sellAssetNetwork } = useGetNetworkQuery(
+    sellToken ?? skipToken
+  )
+
+  const { data: buyAssetNetwork } = useGetNetworkQuery(
+    buyToken ?? skipToken
+  )
+
+  const {
+    data: sellTokenDecimals
+  } = useGetEthTokenDecimalsQuery(
+    sellToken?.coingeckoId === UNKNOWN_TOKEN_COINGECKO_ID &&
+    sellToken?.contractAddress !== ''
+      ? {
+          chainId: sellToken.chainId,
+          contractAddress: sellToken.contractAddress
+       }
+      : skipToken
+  )
+  const {
+    data: sellTokenSymbol
+  } = useGetEthTokenSymbolQuery(
+    sellToken?.coingeckoId === UNKNOWN_TOKEN_COINGECKO_ID &&
+    sellToken?.contractAddress !== ''
+      ? {
+          chainId: sellToken.chainId,
+          contractAddress: sellToken.contractAddress
+        }
+      : skipToken
+  )
+
+  const {
+    data: buyTokenDecimals
+  } = useGetEthTokenDecimalsQuery(
+    buyToken?.coingeckoId === UNKNOWN_TOKEN_COINGECKO_ID &&
+    buyToken?.contractAddress !== ''
+      ? {
+          chainId: buyToken.chainId,
+          contractAddress: buyToken.contractAddress
+       }
+      : skipToken
+  )
+  const {
+    data: buyTokenSymbol
+  } = useGetEthTokenSymbolQuery(
+    buyToken?.coingeckoId === UNKNOWN_TOKEN_COINGECKO_ID &&
+    buyToken?.contractAddress !== ''
+      ? {
+          chainId: buyToken.chainId,
+          contractAddress: buyToken.contractAddress
+        }
+      : skipToken
+  )
+
+  const buyTokenResult = makeToken(
+    buyToken,
+    buyAssetNetwork,
+    buyTokenSymbol,
+    buyTokenDecimals
+  )
+  const sellTokenResult = makeToken(
+    sellToken,
+    sellAssetNetwork,
+    sellTokenSymbol,
+    sellTokenDecimals
+  )
```

Comment:
memoize these objet to prevent re-renders of child components

Reply:
https://github.com/brave/brave-core/pull/20203/commits/e075c8e2829179099b2af490db06926db1b88000